Add option to serialize enums as their names

ydylla opened the pull request:
This implements #207  
An option to serialize enums as their names instead of values.

```python
class Foo(enum.Enum):
    ONE = 1

print(orjson.dumps(Foo.ONE, option=orjson.OPT_ENUM_NAME))
```
Which will print `b'"ONE"'`

Right now the `AutoEnum` test fails because enums created by `enum.auto` are not handled by the enum serialize code.  
I am not sure what the best solution is, since these enums also do not have a `name` attribute.  
Maybe this option/test should just ignore `enum.auto` enums?

ijl commented:
I think the idea is fine. Can you write a test showing the AutoEnum issue? It's also an option to have `OPT_PASSTHROUGH_ENUM` for `default`.

ydylla commented:
This PR already contains a failing AutoEnum Test [here](https://github.com/ijl/orjson/pull/208/files#diff-f28b681350ac72fba6e05691a6cba6b11f7341109e105bbda307af710f37369eR126). But here is a standalone one:
```python
def test_auto_enum_with_name_option(self):
    class AutoEnum(enum.auto):
        A = "a"
    self.assertEqual(orjson.dumps(AutoEnum.A, option=orjson.OPT_ENUM_NAME), b'"A"')
```

The Problem is that enum.auto does not seam to generate real python enums. They are not detected as enums by orjson and also do not have a `name` attribute. `AutoEnum.A.name` does not work.

An `OPT_PASSTHROUGH_ENUM` option would work too. But since the most common use case for that would probably be to serialize them as their names I would preferrer the dedicated `OPT_ENUM_NAME` option, which is very likely faster (I did not test this). Or simply add both 😃

ydylla commented:
Hi @ijl,
any update on this?

I just rebased onto master and added docs to the readme.
In my opinion it works good enough even without support for enum.auto. Users that use auto.enums will probably not use this option anyway because in python these enums do not have a name attribute.

ydylla commented:
On second thought I think the usage of enum.auto here [test_enum.py#L30](https://github.com/ijl/orjson/blob/master/test/test_enum.py#L30) is just wrong. Following the [Python docs](https://docs.python.org/3/library/enum.html#using-automatic-values) it should look like this:
```python
class AutoEnum(enum.Enum):
    A = enum.auto()
```
Which would work just fine with the code in this pull request, since now `isinstance(AutoEnum.A, enum.Enum)` would be `true` and the name attribute exists. On master AutoEnum.A is just a normal str of a normal class called AutoEnum.

ydylla commented:
Hi @ijl
are you still willing to add this feature?
If not please close this pull request, so I can stop annoy you 😃

georgemarshall commented:
Any update as to what is blocking these changes?

ydylla commented:
> update as to what is blocking these changes?

It is ready to be merged. Just waiting for somebody with permission to do so.

gen-xu commented:
It might be good to also add support with typename so it corresponding with the `__str__` of the enum.
i.e. if we have enum like this
```python
class Foo(enum.Enum):
  A = 1
  B = 2
```
`Foo.A` will serialize into `b'Foo.A'`

atellier commented:
Hi @ijl,

Any chance you could merge this PR?

Thanks.

atellier commented:
> I think the idea is fine. Can you write a test showing the AutoEnum issue? It's also an option to have `OPT_PASSTHROUGH_ENUM` for `default`.

That would be ideal - so we can use str() as well instead and let the enum derived class chose.

georgemarshall commented:
> Any chance you could merge this PR?

@atellier at this point I'm of the opinion that @ijl is not interested in merging this. Their lack of engagement in the PR is a very telling sign.

As it stands the current `Enum` implementation in `orjson` is "broken" by diverging from the convention that other libraries such as `SQLAlchemy` use. The convention being that the `name` portion is the unique part, not the `value`.

atellier commented:
> > Any chance you could merge this PR?
> 
> @atellier at this point I'm of the opinion that @ijl is not interested in merging this. Their lack of engagement in the PR is a very telling sign.
> 
> As it stands the current `Enum` implementation in `orjson` is "broken" by diverging from the convention that other libraries such as `SQLAlchemy` use. The convention being that the `name` portion is the unique part, not the `value`.

Agreed, `name` sounds like the senibble option when you want to try standardize between your relational database and other JSON datastores. To be honest, some other libs such as `cattrs` (`unstructure`) use `value` also by default, but at least provide hooks to override (`register_structure_hook` in the case of `cattrs`). A good compromise could be the `OPT_PASSTHROUGH_ENUM` proposal.

testing12331 commented:
Any update on this? I would like to use this feature in my project.

ydylla commented:
@ijl Could you elaborate a bit more what you think is still missing? I would be happy to help.
In my opinion this was merge ready and also just a small change, without any negative effects on general performance.  
But a huge performance improvement for people that need enum names to be serialized.

ydylla commented:
Okay I just noticed your refactoring from yesterday has introduced conflicts.
I have rebased my [enum-names](https://github.com/ydylla/orjson/tree/enum-names) branch again and resolved the conflicts.

FranciscoDA commented:
Hello.

We prefer serializing enums using their names since `MyEnum[mykey]` seems faster than `MyEnum(myval)`, this would allow users to retrieve the enum instance more efficiently and eliminates ambiguities due to repeated values in the enum class. This feature would be very useful.

Are there any plans to merge this PR?

GabrielCappelli commented:
I've just come across an use case where I need to write Enum values to one destination, but names to another, would be pretty useful if orjson supported this.

ydylla commented:
Hi @FranciscoDA & @GabrielCappelli,
sadly ijl does not seam to want this option merged.

But I have just rebased again and also created a release in my fork. You could install the wheels from there:
https://github.com/ydylla/orjson/releases/tag/3.8.1%2Benum

Disclaimer: I do not intend to support the fork except what is necessary for my own use.

RHeynsZa commented:
Eh, i was already hesitant using this package, specifically because the debugging is so bad. But this was the nail in the coffin. 

`OPT_ENUM_NAME` was a great idea @ydylla

swpinkerton commented:
@ijl Would you be open to looking into either `OPT_PASSTHROUGH_ENUM ` or `OPT_ENUM_NAME`?
I would love to use the library, but this issue remains a sticking point.
